main/src/ca/uwaterloo/flix/language/ast/Purity.scala

```diff
@@ -33,4 +33,50 @@ object Purity {
    */
   case object Impure extends Purity
 
+  /**
+    * Returns true if `p` is is a purity that allows algebraic effects.
+    */
+  def canUseAlgebraicEffects(p: Purity): Boolean = p match {
+    case Pure => false
+    case Impure => true
+  }
+
+  /**
+    * Returns the max purity of `p1` and `p2` according to this ordering:
+    * Pure < Impure
+    */
+  def combine(p1: Purity, p2: Purity): Purity = (p1, p2) match {
+    case (Pure, Pure) => Pure
+    case (Pure, Impure) => Impure
+    case (Impure, Pure) => Impure
+    case (Impure, Impure) => Impure
+  }
+
+  /**
+    * Returns the max purity of `p1`, `p2`, and `p3` according to this ordering:
+    * Pure < Impure
+    */
+  def combine(p1: Purity, p2: Purity, p3: Purity): Purity = {
```

Comment:
Can we make a design principle to avoid overloading? It makes errors worse and it's un-Flixlike.

Reply:
We use overloading in other places. But I am ok with going for combine3.

Reply:
Yeah I don't like that we use it other places either :D

Reply:
Burning the midnight oil or up early?

Reply:
Seems likely to be both. A migraine is pain and no sleep. A caffeine pill solves one of those.

Reply:
Sorry to hear that :O